### ISSUE TITLE
Using NoArgsConstructor when an other struct is using it

### ISSUE THREAD

**Issue by jaudiger** [sokomishalov/lombok-rs#5]
Hi again,

First of all, thank you! The changes you made are really appreciated. I'm using your library in one of my source codes, and it's really helpful.

I was continuing to apply it in most of the situations I can. And I encounter a problem when I derive a structure with NoArgsConstructor which contains another structure with the same derivation. Let's take the following example:

```rust
use lombok::NoArgsConstructor;

#[derive(NoArgsConstructor)]
struct A {
    my_var: String,
}

#[derive(NoArgsConstructor)]
struct B {
    my_other_struct: A,
}
```

The compiler is complaining with: `the trait bound A: std::default::Default is not satisfied
required by std::default::Default::default`. One way to fix the issue is to derive A with the trait Default. But I was wondering if `NoArgsConstructor` could reuse the constructor `new_default` to construct its own constructor.

Anyway, one other question, is `new_default` the default naming in Rust for no constructor argument? Or could it be `new` or `default`?

**Comment by NobodyXu** [comment]
I don't think it is possible to call `new_default` without defining yet another trait for `new_default`, since
 - `new_default` is not standardlized name
 - Only way to automatically check whether `new_default` is available is to use a trait (rust lacks function level reflection) and according to [here](https://stackoverflow.com/questions/47709129/how-to-check-whether-a-generic-type-implements-a-generic-trait), only way to check this is to yet implement another trait, which is just too complated.
   
   It is easier to provide default implementation for all types implements `std::default::Default` for `NoArgsConstructor`.

But then, what is the difference between `NewDefault` and `std::default::Default`?
If a struct implements `NoArgsConstructor`, then it is likely it wants `std::default::Default` too, at least that is what I think and what rust linter thinks.

**Comment by jaudiger** [comment]
Using a new trait just would be too much cumbersome here. That's a bit frustrating that Rust does not support yet object reflection.

This situation can be annoying the first time as I described in my initial comment. So, maybe to summarize the situation following your last point in your comment, maybe the only thing to do here would be to inform the persons using `NoArgsConstructor` to also derive the trait Default when they want to use `NoArgsConstructor` with a structure already using it.

**Comment by NobodyXu** [comment]
Or maybe `NoArgsConstructor` should also automatically derive `Default`, since it makes perfect sense and linter of rust gives you a warning if you implement a no arg constructor `fn new() -> type` but do not derive `Default`.

**Comment by jaudiger** [comment]
In that case, it could make sense to automatically derive `Default` when deriving `NoArgsConstructor` if the linter is already complaining when doing the same thing "manually".

**Comment by sokomishalov** [comment]
Hi! Thanks for raising it.
`new_default()` is definitely not a standardized name for this method. The only a made this name - to prevent collision in case of using with `AllArgsConstructor` which already has `new(...)` method and rustc fails to compile due to rust does not support method overloading yet, unfortunately. 
So, from my perspective, there are two completely different ways to solve this issue:

1) 
 - provide new trait `NoArgsConstructor` with the only method `fn new() -> Self`
 - implement it while deriving
 - in case of nested deriving, we should check if this nested one implements either `Default` or `NoArgsConstructor` (by this [hack](https://stackoverflow.com/a/47714313/5843129)) and choose the existing one.
 
 2)
 - implicitly derive `Default` while deriving `NoArgsConstructor`
 - use this [hack/approach/whatever](https://stackoverflow.com/a/42238787/5843129) to provide standardized `fn new() -> Self` method without collision with `AllArgsConstructor`
 
 And I personally prefer the second approach - seems easier, less hacky, and more useful 
 How do you think?

**Comment by NobodyXu** [comment]
I personally would pick the second approach over the first as I believe it is a cleaner approach and also because rust linter seems to require method with no arg contructor anyway.

**Comment by sokomishalov** [comment]
Well, I've released patch release 0.3.1 with fixing nested struct issues via automatically deriving `Default` trait, but haven't fixed the method naming issue yet - there is still the `new_default()` method there for backward compatibility and future experiments to workaround overloading issue, but you are able to use simply `default()` method from `Default` trait
